pex/tools/commands/venv.py

```diff
@@ -61,12 +63,206 @@ def _copytree(
                         link = False
                 shutil.copy(src_entry, dst_entry)
             except OSError as e:
-                if e.errno == errno.EEXIST:
-                    pex_warnings.warn(
-                        "Failed to overwrite {} with {}: {}".format(dst_entry, src_entry, e)
+                if e.errno != errno.EEXIST:
+                    raise e
+
+
+class CollisionError(Exception):
+    """Indicates multiple distributions provided the same file when merging a PEX into a venv."""
+
+
+def populate_venv_with_pex(
+    venv,  # type: Virtualenv
+    pex,  # type: PEX
+    bin_path=BinPath.FALSE,  # type: BinPath.Value
+    python=None,  # type: Optional[str]
+    collisions_ok=True,  # type: bool
+):
+    # type: (...) -> None
+
+    venv_python = python or venv.interpreter.binary
+    venv_bin_dir = os.path.dirname(python) if python else venv.bin_dir
+    venv_dir = os.path.dirname(venv_bin_dir) if python else venv.venv_dir
+
+    # 1. Populate the venv with the PEX contents.
+    provenance = defaultdict(list)
+
+    def record_provenance(src_to_dst):
+        # type: (Iterable[Tuple[str, str]]) -> None
+        for src, dst in src_to_dst:
+            provenance[dst].append(src)
+
+    pex_info = pex.pex_info()
+    if zipfile.is_zipfile(pex.path()):
+        record_provenance(
+            PEXEnvironment.explode_code(
+                pex.path(), pex_info, venv.site_packages_dir, exclude=("__main__.py",)
+            )
+        )
+    else:
+        record_provenance(
+            _copytree(
+                src=pex.path(),
+                dst=venv.site_packages_dir,
+                exclude=(pex_info.internal_cache, pex_builder.BOOTSTRAP_DIR, "__main__.py"),
+            )
+        )
+
+    for dist in pex.activate():
+        record_provenance(
+            _copytree(src=dist.location, dst=venv.site_packages_dir, exclude=("bin",))
+        )
+        dist_bin_dir = os.path.join(dist.location, "bin")
+        if os.path.isdir(dist_bin_dir):
+            record_provenance(_copytree(dist_bin_dir, venv.bin_dir))
+
+    collisions = {dst: srcs for dst, srcs in provenance.items() if len(srcs) > 1}
+    if collisions:
+        message_lines = [
+            "Encountered {collision} building venv at {venv_dir} from {pex}:".format(
+                collision=pluralize(collisions, "collision"), venv_dir=venv_dir, pex=pex.path()
+            )
+        ]
+        for index, (dst, srcs) in enumerate(collisions.items(), start=1):
+            message_lines.append(
+                "{index}. {dst} was provided by:\n\t{srcs}".format(
+                    index=index, dst=dst, srcs="\n\t".join(srcs)
+                )
+            )
+        message = "\n".join(message_lines)
+        if not collisions_ok:
+            raise CollisionError(message)
+        pex_warnings.warn(message)
+
+    # 2. Add a __main__ to the root of the venv for running the venv dir like a loose PEX dir
+    # and a main.py for running as a script.
```

Comment:
Is it worth checking for a collision with `__main__.py`?

Reply:
There can be none. A venv is layed out as:
```
<venv dir>/
  bin/
  lib/
  include/
  pyvenv.cfg
```
The `__main__.py` and the `pex` symlink are dropped in the `<venv dir>` root, not in `lib/.../site-packages` which is the place where collisions could occur.